File: Projeto_4/src/main.cpp
#include <iostream>
#include "ContaBancaria.h"
#include "Transacao.h"

// Implementação da Função Amiga (não pertence a nenhuma das classes)
bool validarTransacao(const ContaBancaria& conta, const Transacao& transacao) {
    // Como é 'friend', pode acessar 'saldo' (protected) e 'valor' (private) diretamente
    if (transacao.valor <= conta.saldo) {
        std::cout << "[AUDITORIA] APROVADA: Transacao de R$" << transacao.valor 
                  << " liberada. Saldo atual: R$" << conta.saldo << ".\n";
        return true;
    } else {
        std::cout << "[AUDITORIA] NEGADA: Tentativa de transacao de R$" << transacao.valor 
                  << " excede o saldo de R$" << conta.saldo << ".\n";
        return false;
    }
}

int main() {
    std::cout << "--- Sistema SecureBank Pro ---\n\n";

    // Criando uma conta
    ContaBancaria conta("Ana Oliveira", "123.456.789-00", 5000.0);
    conta.exibirDados();

    std::cout << "------------------------------\n";

    // Simulando transações
    Transacao t1(2500.0, "30/08/2026");
    Transacao t2(6000.0, "30/08/2026");

    // Validações via Módulo de Auditoria
    std::cout << "Analisando Transacao 1...\n";
    validarTransacao(conta, t1);

    std::cout << "\nAnalisando Transacao 2...\n";
    validarTransacao(conta, t2);

    return 0;
}

File: Projeto_4/main.py
from models.ContaBancaria import ContaBancaria
from models.Transacao import Transacao

def validar_transacao(conta: ContaBancaria, transacao: Transacao) -> bool:
    """
    Simula uma 'friend function'. 
    Acessa diretamente os atributos da conta e da transação sem getters.
    """
    saldo_atual = conta._saldo 
    valor_transacao = transacao._Transacao__valor 
    
    print("\n--- Iniciando Auditoria de Transação ---")
    if valor_transacao <= saldo_atual:
        print("Transação Válida. Valor dentro do limite de saldo.")
        return True
    else:
        print("Transação Inválida! O valor excede o saldo disponível.")
        return False

if __name__ == "__main__":
    conta_cliente = ContaBancaria("Emyliano Carvalho Costa", "123.456.789-10", 2500.00)
    tentativa_transacao = Transacao(3000.00, "02-08-2026")
    transacao_segura = Transacao(450.00, "02-08-2026")

    print("=== Dados da Conta ===")
    conta_cliente.exibirDaddos()

    validar_transacao(conta_cliente, tentativa_transacao)
    validar_transacao(conta_cliente, transacao_segura)